backend/services/implementations/testSessionService.ts

```diff
@@ -314,31 +315,68 @@ class TestSessionService implements ITestSessionService {
   ): Promise<ResultResponseDTO> {
     let resultResponseDTO: ResultResponseDTO;
 
-    // the list of a student's answers with each field being either the
-    // numeric answer (for short answer) or index (for multiple choice)
-    const studentAnswers: (number | null)[] = result.answers;
+    // the list of a student's answers with each field being either the:
+    // - numeric answer (for short answer)
+    // - index (for multiple choice)
+    // - list of indices (for multiple select)
+    // - null (for no answer)
+    const studentAnswers: (number[] | number | null)[][] = result.answers;
 
     let computedScore = 0.0;
-    const computedBreakdown: boolean[] = [];
+    const computedBreakdown: boolean[][] = [];
     let questionsCorrect = 0;
+    let questionsCount = 0;
 
     try {
       const test: TestResponseDTO = await this.testService.getTestById(testId);
-
-      test.questions.forEach((question: Question, i) => {
-        const actualAnswer: number = this.getCorrectAnswer(question);
-
-        if (studentAnswers[i] === actualAnswer) {
-          questionsCorrect += 1;
-          computedBreakdown[i] = true;
-        } else {
-          computedBreakdown[i] = false;
-        }
+      test.questions.forEach((questionComponents: QuestionComponent[], i) => {
+        const computedBreakdownByQuestion: boolean[] = [];
+        questionComponents.forEach((questionComponent: QuestionComponent) => {
+          const { type } = questionComponent;
+          if (
+            type === QuestionComponentType.MULTIPLE_CHOICE ||
+            type === QuestionComponentType.SHORT_ANSWER
+          ) {
+            const actualAnswer: number = this.getCorrectAnswer(
+              questionComponent,
+            );
+            const studentAnswer = studentAnswers[i][questionsCount] as
+              | number
+              | null;
+
+            if (studentAnswer === actualAnswer) {
+              questionsCorrect += 1;
+              computedBreakdownByQuestion.push(true);
+            } else {
+              computedBreakdownByQuestion.push(false);
+            }
+            questionsCount += 1;
+          } else if (type === QuestionComponentType.MULTI_SELECT) {
+            const actualAnswers: number[] = this.getCorrectAnswers(
+              questionComponent,
+            );
+            const studentAnswer = studentAnswers[i][questionsCount] as
+              | number[]
+              | null;
+
+            if (
+              studentAnswer?.length === actualAnswers.length &&
+              studentAnswer.every((val, idx) => val === actualAnswers[idx])
+            ) {
+              questionsCorrect += 1;
+              computedBreakdownByQuestion.push(true);
+            } else {
+              computedBreakdownByQuestion.push(false);
+            }
+            questionsCount += 1;
+          }
+        });
+        computedBreakdown.push(computedBreakdownByQuestion);
```

Comment:
nit: for the different cases in the for loop, maybe it makes sense to just determine whether the student answer is correct i.e ( `true` / `false`) inside the `if` statement? And then the rest of the logic could be extracted outside the if (i.e logic in line 367, 369, 371, etc.) to avoid duplication?

or does that not work because of the other `QuestionComponentTypes`?